### PR TITLE
Move regexes to source/common/regular-expressions

### PR DESCRIPTION
**Pull request by Kangie**
## Description

Moved all of the regexes that I could find to source/common/regular-expressions
## Changes

Regexes removed from source files in favour of calls to functions in source/common/regular-expressions
## Additional information

I'm pretty sure that I got all of the regexes, but I'm 100% sure that I missed something with a non-standard name.

I've done my best to keep the names consistent, but where the same RE was called within two different functions I've made a call. 

List RE functions have all been given consistent names.

At some point, it would be a good idea to alphabetise the list of functions in source/common/regular-expressions

I've tested as many of the functions as I can; various syntax highlighting / markup things all seem to be working:

- Lists
- Images
- URLs
- Footnotes
- References

Let me know if you have any feedback; this should make all of the Regexes easier to maintain.

Tested on: <!-- OS including version ->
Gentoo Linux.

### CODE REVIEW

File: source/common/regular-expressions.js [hunk 1]
```diff
@@ -14,6 +14,190 @@
...
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown Headings.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getHeadingRE': function () {
+    return RegExp(
+      /(#+)\s+/.source
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:153]: Don't we need to assure start-of-line with ^?

File: source/common/regular-expressions.js [hunk 2]
```diff
@@ -14,6 +14,190 @@
...
+    )
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown highlighting (???).
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:158]: To your question marks: Yeah, this is non-standard syntax that I've added based on a request. It's supported almost nowhere, but the intention is to get certain strips of Markdown displayed textmarker-like (I think Zettlr's themes use yellow background for this)

File: source/common/regular-expressions.js [hunk 3]
```diff
@@ -14,6 +14,190 @@
...
+  /**
+   * Returns a regular expression that matches file IDs as in the settings
+   *
+   * @param   {boolean}  multiline  Whether to match multiline
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getIDRE': function (multiline = false) {
+    let flag = (multiline) ? 'm' : ''
+    let idRegExpString = global.config.get('zkn.idRE')
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:177]: Did you make sure this function doesn't get called before the configuration provider is loaded? (Just asking because I'm always afraid of such race conditions :D)

File: source/common/regular-expressions.js [hunk 4]
```diff
@@ -14,6 +14,190 @@
...
+  },
+
+  /**
+   * Returns a regular expression that matches inline frames.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getIframeRE': function () {
+    return RegExp(
+      /^<iframe.*?>.*?<\/iframe>$/.source,
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:197]: Note to myself: We definitely need a better way to get iFrames, because right now it's still that users can actually open a door to malware, because this will just load *any* iFrame, not just "good" ones (e.g. YouTube, Vimeo, etc.)

File: source/common/regular-expressions.js [hunk 5]
```diff
@@ -43,6 +228,117 @@ module.exports = {
...
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getListTaskListRE': function () {
+    return RegExp(
+      /^(\s*)(- \[[x ]\])(\s*)/.source
+    )
+  },
+
+  /**
+   * Returns a regular expression that matches unordered Markdown lists.
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:299]: * and includes a capturing group to capture the list tokens, right?

File: source/common/regular-expressions.js [hunk 6]
```diff
@@ -43,6 +228,117 @@ module.exports = {
...
+      /[*+-]\s/.source
+    )
+  },
+
+  /**
+   * Returns a regular expression that matches unordered MarkDown lists.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getListUnorderedCMRE': function () {
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:325]: CM?

File: source/common/regular-expressions.js [hunk 7]
```diff
@@ -43,6 +228,117 @@ module.exports = {
...
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown files.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getMarkDownFileRE': function () {
+    return RegExp(
+      /.+\.(?:md|markdown|txt)$/.source,
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:338]: RMarkdown is missing here (rmd). I've let that one out of many regular expressions until now because I read that Adobe uses .rmd as some binary file format as well, but I'm close to simplifying the code to just allow RMarkdown by default (which would remove the corresponding setting as well) …

File: source/common/regular-expressions.js [hunk 8]
```diff
@@ -57,25 +353,71 @@ module.exports = {
...
-      idRegExpString = `(${idRegExpString})`
-    }
+  'getTableRE': function () {
+    return RegExp(
+      /^\|.+\|$/.source,
+      'i')
+  },
 
+  /**
+   * Returns a regular expression that matches inline frames.
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:369]: Are you sure that's what this RegExp does? :D

File: source/common/regular-expressions.js [hunk 9]
```diff
@@ -57,25 +353,71 @@ module.exports = {
...
...+)\]\((.+?)\)|(((?:(?:aaas?|about|acap|adiumxtra|af[ps]|aim|apt|attachment|aw|beshare|bitcoin|bolo|callto|cap|chrome(?:-extension)?|cid|coap|com-eventbrite-attendee|content|crid|cvs|data|dav|dict|dlna-(?:playcontainer|playsingle)|dns|doi|dtn|dvb|ed2k|facetime|feed|file|finger|fish|ftp|geo|gg|git|gizmoproject|go|gopher|gtalk|h323|hcp|https?|iax|icap|icon|im|imap|info|ipn|ipp|irc[6s]?|iris(?:\.beep|\.lwz|\.xpc|\.xpcs)?|itms|jar|javascript|jms|keyparc|lastfm|ldaps?|magnet|mailto|maps|market|message|mid|mms|ms-help|msnim|msrps?|mtqp|mumble|mupdate|mvn|news|nfs|nih?|nntp|notes|oid|opaquelocktoken|palm|paparazzi|platform|pop|pres|proxy|psyc|query|res(?:ource)?|rmi|rsync|rtmp|rtsp|secondlife|service|session|sftp|sgn|shttp|sieve|sips?|skype|sm[bs]|snmp|soap\.beeps?|soldat|spotify|ssh|steam|svn|tag|teamspeak|tel(?:net)?|tftp|things|thismessage|tip|tn3270|tv|udp|unreal|urn|ut2004|vemmi|ventrilo|view-source|webcal|wss?|wtai|wyciwyg|xcon(?:-userid)?|xfire|xmlrpc\.beeps?|xmpp|xri|ymsgr|z39\.50[rs]?):(?:\/{1,3}|[a-z0-9%])|www\d{0,3}[.]|[a-z0-9.-]+[.][a-z]{2,4}\/)(?:[^\s()<>]|\([^\s()<>]*\))+(?:\([^\s()<>]*\)|[^\s`*!()[\]{};:'".,<>?«»“”‘’])))|([a-z0-9.\-_+]+?@[a-z0-9.\-_+]+\.[a-z]{2,7})$/.source,
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:397]: Note to myself: We definitely should get to simplifying this monster 👀

File: source/common/regular-expressions.js [hunk 10]
```diff
@@ -57,25 +353,71 @@ module.exports = {
...
+  },
+
+  /**
+   * Returns a regular expression that matches Zettelkasten IDs.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getZknTagRE': function () {
+    return RegExp(
+      /##?[^\s,.:;…!?"'`»«“”‘’—–@$%&*#^+~÷\\/|<=>[\](){}]+#?/.source,
```
Comment (nathanlesage) [inline source/common/regular-expressions.js:419]: I remember that I've added a negative lookbehind to make sure the hashtag is either on the beginning of a line or preceded by a space, but then I don't know where this regexp is being used …

In general: It's a very good idea to collect all REs here, because there's so many potential issues you can spot with this!

File: source/renderer/modules/markdown-editor/plugins/render-citations.js [hunk 11]
```diff
@@ -45,7 +46,7 @@
   // blocks, second alternative are the simple @ID-things, both recognised by
   // Pandoc citeproc.
   // citationRE is taken from the Citr library (the extraction regex)
```
Comment (nathanlesage) [inline source/renderer/modules/markdown-editor/plugins/render-citations.js:48]: It might make sense to add this comment above the getCitationRE-func in the regexp-file, so that it's clear where it comes from (and, whenever I should change this RE, I know that I must update this as well)

**Review by nathanlesage — APPROVED** [review]
Amazing work, thank you so much! In general I think this is exactly how it needs to be :)

Just one question: Would you mind double-checking each affected file that the functionality still works out? (Please also feel free to add corresponding GUI-checks to the scripts/test-gui - folder)

If everything works out, then we can merge it :) This code is platform-independent so it suffices if it works out on Windows!

File: source/common/regular-expressions.js (hunk 1 above)
Comment (Kangie) [reply source/common/regular-expressions.js:153]: I'll validate this against the current source tomorrow, but I did my best to make exact copies of the existing REs. Better to make changes once I'm sure I haven't broken anything!

File: source/common/regular-expressions.js (hunk 2 above)
Comment (Kangie) [reply source/common/regular-expressions.js:158]: Fantastic. I'll remember that next time I want you to implement something that's not in common mark :P

I'll update the docs!

File: source/common/regular-expressions.js (hunk 3 above)
Comment (Kangie) [reply source/common/regular-expressions.js:177]: Pretty sure that one was already in this file, but I'll take a look.

**Comment by Kangie** [comment]
> Amazing work, thank you so much! In general I think this is exactly how it needs to be :)

I was just following your lead; at some point you'd put a few common REs here. This way they're all in one place when I need to steal one for working with MarkDown!

Thanks for the feedback! I haven't tested everything yet, but everything that I've tested so far has worked. I'll continue to work on it and implement any further feedback. 

I may enable some additional linting rules with your approval - enforce function alphabetisation in the long term, and make it a warning for now.

File: source/common/regular-expressions.js (hunk 8 above)
Comment (Kangie) [reply source/common/regular-expressions.js:369]: Yes. 100%. The function name is merely obfuscation.

File: source/common/regular-expressions.js (hunk 6 above)
Comment (Kangie) [reply source/common/regular-expressions.js:325]: Codemirror. 
This RE is only used by codemirror stuff. It could probably be consolidated at some point, as there's another RE that does pretty much the same thing, but I don't want to change things until I know this hasn't broken anything.

**Comment by nathanlesage** [comment]
Perfect! I know testing is cumbersome as you have to trigger all functions at least once, so I hope you're not mad if I want you to do the testing this time :D

As for alphabetisation rules: I'm not extremely sure if that's useful in any case: for instance the service providers all feature a shutdown-function and I like them to be at the same position under the constructor, which wouldn't work in that case. But then, I never tried out something else so I might very well like it :D

File: source/common/regular-expressions.js (hunk 1 above)
Comment (Kangie) [reply source/common/regular-expressions.js:153]: The current regex doesn't do that, so I won't change this for fear of unintended consequences, unless you'd prefer that I update it in this PR?

File: source/common/regular-expressions.js (hunk 3 above)
Comment (Kangie) [reply source/common/regular-expressions.js:177]: This code is unchanged.

File: source/common/regular-expressions.js (hunk 7 above)
Comment (Kangie) [reply source/common/regular-expressions.js:338]: Won't change for now, unless you insist, same as above. :)

File: source/common/regular-expressions.js (hunk 9 above)
Comment (Kangie) [reply source/common/regular-expressions.js:397]: I like needing to scroll for five minutes to parse a regex.

File: source/common/regular-expressions.js (hunk 10 above)
Comment (Kangie) [reply source/common/regular-expressions.js:419]: `.../markdown-editor/modes/spellchecker.js` and `.../markdown-editor/modes/markdown-zkn.js` used to have independent implementations of this regex.

**Comment by Kangie** [comment]
Tests: 

| Module | Result |
| ----------- | --------- |
| make-valid-uri | http/https/file URIs work |
| dir-project-export | Unable to test (my pandoc install is broken pending some dependencies from distro. ) |
| footnotes | Success! Tooltips! |
| autocomplete | Success! Syntax highlighting for code works. This syntax isn't implemented though: {.tex .numberlines} |
| markdown-zkn |  I don't use/understand this functionality well enough to test properly. All the REs that it calls do seem to work though|
| spellchecker | Unknown words within code/footnote/zkn ignored |
| continuelist | Fail. Then pass. :D |
| indentlist | Pass. Numbering doesn't restart though - did it ever? I Haven't written a new ordered list in Zettlr in a couple of months. |
| markdown-shortcuts | Pass. Everything seems to highlight successfully. |
| render-citation | Short form works - I assume that the long form works too. |
| render-iframes | Not sure how this one is supposed to work - have never used an iframe in Zettlr. Can't make anything happen, will test against develop tomorrow. |
| render-math | multiline works, inline regex matches but rendering is broken.  |
| render-tables | Tables seem to render in WYSIWYG. Headings are properly bold. |
| render-tasks | Regex matches, but rendering is broken | 
| wysiwyg | All of the expected formatting seems to work | 

So to merge, Validate / Fix: 

- [ ] Task List
- [ ] Inline Maths
- [ ] iframes

Pretty happy with that overall. :)

I probably won't write tests at this point, but with everything consolidated that's a lot easier than it was before.

**Comment by Kangie** [comment]
Iframes behave the same way on develop, so I'm calling that a pass. Task lists still aren't working... but everything else is. The regex is unchanged. Why?!

I'll look into it tomorrow hopefully. If you can spot the obvious error that I've made let me know!

**Comment by Kangie** [comment]
OK, Rebased to fix the merge conflict and tidy up the history. Should be good to go. 

Lint error seems to have been introduced in the tables space, but not in any module that I've touched.

**Comment by nathanlesage** [comment]
Yeah, linter error was my fault, but I've fixed it just now. I'll be fetching your branch later today and just test it out quickly, but I don't think that all too much should go wrong! That's, after all, what the `test-gui` is for, maybe I'll find some edge cases validating new contents to those files …

> I think I've spent 16 of the past 24 hours working with JS for one project or another...

Reminds me very much of myself …

**Comment by Kangie** [comment]
Rebased to fix merge conflict. Looks like export-project-dir is now typescript (along with a lot of other commands!). 

Fixed up the RE import in export-project-dir, and the linting in dir-delete.

File: source/common/regular-expressions.js (hunk 1 above)
Comment (nathanlesage) [reply source/common/regular-expressions.js:174]: No, no, if that breaks things, better not. I concur with you to better play it safe, especially with these big PRs!

File: source/common/regular-expressions.js (hunk 7 above)
Comment (nathanlesage) [reply source/common/regular-expressions.js:387]: Yes, you're right! Let's see that as a note to myself, and probably you, if you get bored on the holidays 😆

File: source/common/regular-expressions.js (hunk 10 above)
Comment (nathanlesage) [reply source/common/regular-expressions.js:477]: That's bad – did you merge them? (Because, after all, both are modes, so both must work on exactly the same basis)

**Comment by nathanlesage** [comment]
> Rebased to fix merge conflict. Looks like export-project-dir is now typescript (along with a lot of other commands!).

Actually, all. I just got the motivation to do that work, and right now I'm happy to having pulled through with this. Migrating to TypeScript is cumbersome as hell. But, alas, I found a few undiscovered bugs, so I count that as a success.

I was about to comment here to tell you that I stumbled upon a regex within a command, but you're fast, I'm really impressed! :)

**Comment by nathanlesage** [comment]
Sooooooo, just pulled your PR and went through the functionality of the files where you've removed the regexps from, and I couldn't find any error, so merging!

You're the best ❤️
